bootstrap/src/basis/file.ml

```diff
@@ -130,49 +130,76 @@ let close t =
+            ~f:(fun _ -> Byte.kv 0L))
+          | Some buffer -> Bytes.Slice.length buffer, buffer
+        end
+      | Some n -> begin
+          match buffer with
+          | None -> n, Bytes.Slice.init (Array.init (0L =:< n) ~f:(fun _ -> Byte.kv 0L))
+          | Some buffer -> (Uns.min n (Bytes.Slice.length buffer)), buffer
+        end
+    end in
+    let inner = finalize_user_data (submit_inner n file) in
+    { inner; buffer }
 
-(* external read_inner: !&bytes -> t >os-> int *)
-external read_inner: Stdlib.Bytes.t -> t -> sint = "hm_basis_file_read_inner"
+  external complete_inner: Stdlib.Bytes.t -> inner -> sint =
+    "hm_basis_file_read_complete_inner"
 
-let read_impl buffer t =
-  let base = Bytes.(Cursor.index (Slice.base buffer)) in
-  let bytes = Stdlib.Bytes.create (Int64.to_int (Bytes.Slice.length buffer)) in
-  let value = read_inner bytes t in
-  match Sint.(value < kv 0L) with
-  | true -> Error (Uns.bits_of_sint Sint.(neg value))
-  | false -> begin
-      let range = (base =:< (base + (Uns.bits_of_sint value))) in
-      let container = Bytes.Slice.container buffer in
-      Range.iter range ~f:(fun i ->
-        Array.set_inplace i (U8.of_char (Stdlib.Bytes.get bytes (Int64.to_int (i - base))))
-          container
-      );
-      Ok (Bytes.Slice.init ~range (Bytes.Slice.container buffer))
-    end
+  let complete t =
+    let base = Bytes.(Cursor.index (Slice.base t.buffer)) in
+    let bytes = Stdlib.Bytes.create (Int64.to_int (Bytes.Slice.length t.buffer)) in
+    let value = complete_inner bytes t.inner in
+    match Sint.(value < kv 0L) with
+    | true -> Error (Uns.bits_of_sint Sint.(neg value))
+    | false -> begin
+        let range = (base =:< (base + (Uns.bits_of_sint value))) in
+        let container = Bytes.Slice.container t.buffer in
+        Range.iter range ~f:(fun i ->
+          Array.set_inplace i (U8.of_char (Stdlib.Bytes.get bytes (Int64.to_int (i - base))))
+            container
+        );
+        Ok (Bytes.Slice.init ~range (Bytes.Slice.container t.buffer))
+      end
+
+  let complete_hlt t =
+    match complete t with
+    | Ok buffer -> buffer
+    | Error error -> halt (Error.to_string error)
+
+end
 
-let read ?(n=read_n) t =
-  let bytes = Array.init (0L =:< n) ~f:(fun _ -> Byte.kv 0L) in
-  let buffer = Bytes.Slice.init bytes in
-  read_impl buffer t
+let read ?n t =
+  Read.(submit ?n t |> complete)
 
 let read_hlt ?n t =
-  match read ?n t with
-  | Ok buffer -> buffer
-  | Error error -> let _ = close_hlt t in halt (Error.to_string error)
+  Read.(submit ?n t |> complete_hlt)
 
 let read_into buffer t =
-  match read_impl buffer t with
+  match Read.(submit ~buffer t |> complete) with
   | Ok _ -> None
   | Error error -> Some error
 
 let read_into_hlt buffer t =
-  match read_into buffer t with
-  | None -> ()
-  | Some error -> begin
-      let _ = close_hlt t in
-      halt (Error.to_string error)
-    end
+  let _ = Read.(submit ~buffer t |> complete_hlt) in
+  ()
```

Comment:
I think the `read_into` functions are somewhat misguided. I'd originally convinced myself they were needed so that Hemlock would see the `read` function as having no mutation effects and `read_into` having mutation effects. If we were to just have a single `read` function that took an optional input buffer, we could mutate an input, but I don't think that's such a bad problem to warrant these additional functions.